odbc/src/api/functions.rs

```diff
@@ -1279,73 +1274,211 @@ pub unsafe extern "C" fn SQLFetch(statement_handle: HStmt) -> SqlReturn {
+                mongo_handle,
+                move_to_next_result.as_ref().unwrap_err().clone(),
+                function_name.to_string()
+            );
+
+            if !row_status_buffer.is_null() {
+                *row_status_buffer = RowStatus::SQL_ROW_ERROR as USmallInt;
+            }
+
+            if has_rows_fetched_buffer {
+                *stmt.attributes.write().unwrap().rows_fetched_ptr += 1;
+            }
+
+            row_error_count += 1;
         }
+    }
+
+    // Only return ERROR if every row causes an error.
+    if row_error_count == rowset_size {
+        SqlReturn::ERROR
+    } else if !global_warnings_opt.is_empty()
+        || encountered_success_with_info
+        || row_error_count > 0
+    {
+        SqlReturn::SUCCESS_WITH_INFO
     } else {
-        add_diag_with_function!(
-            mongo_handle,
-            move_to_next_result.as_ref().unwrap_err().clone(),
-            function_name.to_string()
+        SqlReturn::SUCCESS
+    }
+}
+
+unsafe fn sql_fetch_column_binding_helper(
+    statement_handle: HStmt,
+    index: ULen,
+    row_status_buffer: *mut USmallInt,
+    bound_cols: &HashMap<USmallInt, BoundColInfo>,
+    function_name: &str,
+) -> (bool, bool) {
+    let mongo_handle_for_sql_get_data_helper = MongoHandleRef::from(statement_handle);
+
+    let mut encountered_error_during_col_binding = false;
+    let mut encountered_success_with_info_during_col_binding = false;
+
+    for (col, bound_col_info) in bound_cols.iter() {
+        // Set target_buffer to the correct buffer in the array of buffers
+        let target_buffer = (bound_col_info.target_buffer as ULen
+            + (index * (bound_col_info.buffer_length as ULen)))
+            as Pointer;
+
+        // Set length/indicator buffer to the correct buffer in the array of buffers
+        let len_ind_buffer =
+            (bound_col_info.length_or_indicator as ULen + (index * size_of::<isize>())) as *mut Len;
+
+        let sql_return = sql_get_data_helper(
+            mongo_handle_for_sql_get_data_helper,
+            *col,
+            FromPrimitive::from_i16(bound_col_info.target_type).unwrap(), // this conversion is checked in SQLBindCol, so it is guaranteed to work here.
+            target_buffer,
+            bound_col_info.buffer_length,
+            len_ind_buffer,
+            function_name,
         );
-        // An error happened
-        SqlReturn::ERROR
+
+        match sql_return {
+            SqlReturn::ERROR => {
+                encountered_error_during_col_binding = true;
+            }
+            SqlReturn::SUCCESS_WITH_INFO => {
+                encountered_success_with_info_during_col_binding = true;
+            }
+            _ => {}
+        }
     }
+
+    if !row_status_buffer.is_null() {
+        // It's possible that one binding causes SUCCESS_WITH_INFO and another causes ERROR, so we need to check for an error first
+        // and ignore SUCCESS_WITH_INFO if it also occurs.
```

Comment:
It's not actually specified anywhere what to do if both happen, but I assume this is correct. I would think ERROR would take precedence over SUCCESS_WITH_INFO.